src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -84,25 +100,75 @@ private StreamingClientProvider() {
    * @return A streaming client
    */
   public SnowflakeStreamingIngestClient getClient(Map<String, String> connectorConfig) {
+    SnowflakeStreamingIngestClient resultClient = null;
+
     if (Boolean.parseBoolean(
         connectorConfig.getOrDefault(
             SnowflakeSinkConnectorConfig.ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG,
             Boolean.toString(ENABLE_STREAMING_CLIENT_OPTIMIZATION_DEFAULT)))) {
-      LOGGER.info(
-          "Streaming client optimization is enabled, returning the existing streaming client if"
-              + " valid");
-      this.providerLock.lock();
-      // recreate streaming client if needed
-      if (!StreamingClientHandler.isClientValid(this.parameterEnabledClient)) {
-        LOGGER.error("Current streaming client is invalid, recreating client");
-        this.parameterEnabledClient = this.streamingClientHandler.createClient(connectorConfig);
+      LOGGER.debug(
+          "Streaming client optimization is enabled per worker node. Reusing valid clients when"
+              + " possible");
+
+      try {
+        this.providerLock.lock();
+
+        // remove invalid clients
+        List<Map.Entry<Properties, SnowflakeStreamingIngestClient>> invalidClientProps =
+            this.registeredClientMap.entrySet().stream()
+                .filter(entry -> !StreamingClientHandler.isClientValid(entry.getValue()))
+                .collect(Collectors.toList());
+        if (!invalidClientProps.isEmpty()) {
+          List<String> invalidClientNames = new ArrayList<>();
+
+          for (Map.Entry<Properties, SnowflakeStreamingIngestClient> entry : invalidClientProps) {
+            SnowflakeStreamingIngestClient invalidClient = entry.getValue();
+            invalidClientNames.add(
+                invalidClient != null
+                        && invalidClient.getName() != null
+                        && !invalidClient.getName().isEmpty()
+                    ? invalidClient.getName()
+                    : "noClientNameFound");
+            this.registeredClientMap.remove(entry.getKey());
+          }
+
+          LOGGER.error(
+              "Found and removed {} invalid clients: {}",
+              invalidClientProps.size(),
+              invalidClientNames.toString());
+        }
+
+        // look for client corresponding to the input properties or create new client
+        Properties inputProps = StreamingClientHandler.getClientProperties(connectorConfig);
+        if (this.registeredClientMap.containsKey(inputProps)) {
```

Comment:
What's this mean here by using all the properties as the key? Like if one connector has buffer flush time to be 1s and the other one is 10s, then a separate client is used? But why it can't use the same client?

Reply:
This would be fixed by recent change from @sfc-gh-rcheng

Reply:
yup! discussed in this [thread](https://github.com/snowflakedb/snowflake-kafka-connector/pull/744#discussion_r1399966870) and offline. thanks for the important call out